runscanner/src/main/java/uk/ac/bbsrc/tgac/miso/runscanner/RunProcessor.java

```diff
@@ -0,0 +1,71 @@
+package uk.ac.bbsrc.tgac.miso.runscanner;
+
+import java.io.File;
+import java.util.Arrays;
+import java.util.TimeZone;
+import java.util.function.Supplier;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.prometheus.client.Counter;
+import io.prometheus.client.Histogram;
+import uk.ac.bbsrc.tgac.miso.core.data.type.PlatformType;
+import uk.ac.bbsrc.tgac.miso.core.util.LimsUtils;
+import uk.ac.bbsrc.tgac.miso.dto.NotificationDto;
+
+public abstract class RunProcessor {
+  private static final Counter errors = Counter.build().name("miso_runscanner_errors").help("The number of bad directories encountered.")
+      .labelNames("platform").register();
+
+  public static final Iterable<RunProcessor> INSTANCES = Arrays.asList();
+
+  private static Logger log = LoggerFactory.getLogger(RunProcessor.class);
+
+  private static final Histogram processTime = Histogram.build().buckets(1, 5, 10, 30, 60, 300, 600, 3600)
+      .name("miso_runscanner_directory_process_time").help("Time to process a run directories in seconds.")
+      .labelNames("platform", "instrument").register();
+
+  private final String name;
+
+  private final PlatformType platformType;
+
+  public RunProcessor(PlatformType platformType, String name) {
+    super();
+    this.platformType = platformType;
+    this.name = name;
+  }
+
+  public final Supplier<NotificationDto> apply(File runDirectory, TimeZone tz) {
+    return () -> {
+      long startTime = System.nanoTime();
+      String instrumentName = "unknown";
+      NotificationDto dto = null;
+      try {
+        dto = process(runDirectory, tz);
+        if (!LimsUtils.isStringBlankOrNull(dto.getSequencerName())) {
+          instrumentName = dto.getSequencerName();
+        }
+      } catch (Exception e) {
+        log.error("Failed to process run: " + runDirectory.getPath(), e);
```

Comment:
I think of it this way:
* the processor's job is to scan the run folder and report what it finds. This could be a proper run, or an error
* the scheduler's job is to manage those results

A clear separation of responsibilities would mean that the processor shouldn't decide what gets done with an exception that it can't properly handle itself